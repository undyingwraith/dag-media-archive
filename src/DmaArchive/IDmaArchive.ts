import CID from 'cids';
import {IArchive} from '../types';

/**
 *
 */
export interface IDmaArchive {
	/**
	 *
	 */
	fetch(): Promise<IArchive>

	/**
	 *
	 * @param archive
	 */
	set(archive: IArchive): Promise<CID>

	/**
	 *
	 * @param cid
	 */
	get(cid: CID): Promise<any>

	/**
	 *
	 * @param data
	 */
	put(data: any): Promise<CID>

	/**
	 *
	 * @param path
	 */
	resolve(path: string): Promise<any>

	/**
	 * sets data at specified path in the archive
	 * @param path of where to put the data
	 * @param data to insert at specified path
	 * @returns Promise<CID> cid of the updated (archive can be ignored)
	 */
	update(path: string, data: any): Promise<CID>

	/**
	 *
	 */
	export(): Promise<CID>
}
