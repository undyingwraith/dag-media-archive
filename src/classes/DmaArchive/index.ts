export * from './DmaArchive'
export * from './IDmaArchive'
