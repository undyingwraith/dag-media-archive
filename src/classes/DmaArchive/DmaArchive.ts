import CID from 'cids';
import {IArchive} from '../../types';
import {IDmaArchive} from './IDmaArchive';

export class DmaArchive implements IDmaArchive {
	constructor(private node: any, private current?: CID) {

	}

	fetch(): Promise<IArchive> {
		if (!this.current) return Promise.reject('not initialized');
		return new Promise((resolve, reject) => {
			this.get(this.current!).then((obj: IArchive) => {
				resolve(obj);
			}).catch(reject);
		});
	}

	get(cid: CID): Promise<any> {
		return new Promise((resolve, reject) => {
			this.node.dag.get(cid).then((obj: { value: any }) => {
				resolve(obj.value);
			}).catch(reject);
		});
	}

	set(archive: IArchive): Promise<CID> {
		return new Promise((resolve, reject) => {
			this.put(archive).then((cid: CID) => {
				this.current = cid;
				resolve(cid);
			}).catch(reject);
		});
	}

	put(data: any): Promise<CID> {
		return new Promise((resolve, reject) => {
			this.node.dag.put(data, {format: 'dag-cbor', hashAlg: 'sha3-512'}).then((cid: CID) => {
				resolve(cid);
			}).catch(reject);
		});
	}

	resolve(path: string): Promise<any> {
		if (!this.current) return Promise.reject('not initialized');
		if (path.substr(0, 1) != '/') return Promise.reject('path must start with a "/"');
		let remainderPath: string;
		return new Promise((resolve, reject) => {
			this.node.dag.resolve(`${this.current!.toV1()}${path}`)
				.then((res: IResolveResult) => {
					remainderPath = res.remainderPath;
					return this.get(res.cid);
				})
				.then((data: any) => {
					if (remainderPath !== '') {
						let ls = remainderPath.lastIndexOf('/');
						if (ls === -1) {
							resolve(data[remainderPath]);
						} else {
							//TODO
							reject(remainderPath);
						}
					} else {
						resolve(data);
					}
				})
				.catch(reject);
		});
	}

	update(path: string, data: CID | any): Promise<CID> {
		if (!this.current) return Promise.reject('not initialized');
		if (path.substr(0, 1) != '/') return Promise.reject('path must start with a "/"');
		return new Promise(async (resolve, reject) => {
			if (data instanceof CID) {
				this.updateNext(this.current!, path, data, resolve, reject);
			} else {
				this.put(data).then((cid: CID) => {
					this.updateNext(this.current!, path, cid, resolve, reject);
				}).catch(reject);
			}
		});
	}

	private updateNext(root: CID, path: string, current: CID, resolve: (cid: CID) => void, reject: (e: any) => void) {
		let name = path.substr(path.lastIndexOf('/') + 1);
		let remainder = path.substr(0, path.lastIndexOf('/'));
		let remainderPath: string;
		this.node.dag.resolve(`${root.toV1()}${remainder}`)
			.then((res: IResolveResult) => {
				remainderPath = res.remainderPath;
				return this.get(res.cid);
			})
			.then((node: any) => {
				node[name] = current;
				return this.put(node);
			})
			.then((cid: CID) => {
				if (remainder === '') {
					console.log('update', cid);
					resolve(cid);
				} else {
					console.log('remainder', remainder);
					this.updateNext(root, remainder, cid, resolve, reject);
				}
			})
			.catch(reject);
	}

	export(): Promise<CID> {
		if (!this.current) return Promise.reject('not initialized');
		return Promise.resolve(this.current);
	}
}

interface IResolveResult {
	remainderPath: string
	cid: CID
}
