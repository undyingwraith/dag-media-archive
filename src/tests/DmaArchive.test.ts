import CID from 'cids';
import Ipfs from 'ipfs-http-client';
import {DmaArchive} from '../DmaArchive';
import {CONSTANTS} from './const';

const node = Ipfs.create();


describe('DmaArchive', () => {
	it('put works', async () => {
		const store = new DmaArchive(node, CONSTANTS.baseCid);
		await expect(store.put({})).resolves.toStrictEqual(new CID('bafyriqdqjbhwb2nfkbchrg65ckfpwjxawewjccyneooriyma4tdoiv3doramskqyam764vpoxakku6ldeuam6o62kwozpucom6eevm4ibajzi'));
	});

	it('get works', async () => {
		const store = new DmaArchive(node, CONSTANTS.baseCid);
		await expect(store.get(new CID('bafyriqdqjbhwb2nfkbchrg65ckfpwjxawewjccyneooriyma4tdoiv3doramskqyam764vpoxakku6ldeuam6o62kwozpucom6eevm4ibajzi'))).resolves.toStrictEqual({});
	});

	it('resolve works', async () => {
		const store = new DmaArchive(node, CONSTANTS.baseCidWTwo);
		await expect(store.resolve('/media/ebb941ae-419e-4848-8664-57597ddb214e').then(value => {
			console.log(value);
			return value;
		})).resolves.toMatchSnapshot();
	});

	it('update data works', async () => {
		const store = new DmaArchive(node, CONSTANTS.baseCidWTwo);
		await expect(store.update('/media/ebb941ae-419e-4848-8664-57597ddb214e/source', {hello: 'world'})).resolves.toStrictEqual(new CID('bafyriqbg6un5nazxiege7t5awfgng7ytwaees4vza5pw2knx3as377kusgabwv7tnhglgvdmjjdalqyyc2vwgsrgfghix3i5cnasjqtoxw3ma'));
	});

	it('update cid works', async () => {
		const store = new DmaArchive(node, CONSTANTS.baseCidWTwo);
		await expect(store.update(
			'/media/ebb941ae-419e-4848-8664-57597ddb214e/source',
			new CID('bafybeihe7kldclv7zde2guf4pp2ec2tr4pynegh4gr4m57uwgsptnbjfjm',
			)))
			.resolves.toStrictEqual(new CID('bafyriqhts76doruqnnzt2ikwhfq4jyigy2ldqf5lwauwgksu2p6pgfitxy3q5sr3tb2znjwyqqb5tu4pyng2bwtiyuj4wjtezmtfvu4vlwvzq'));
	});

	it('update \'path must start with a "/"\'', async () => {
		const store = new DmaArchive(node, CONSTANTS.baseCid);
		await expect(store.update('a/b/c', {hello: 'world'})).rejects.toBe('path must start with a "/"');
	});

	it('resolve \'path must start with a "/"\'', async () => {
		const store = new DmaArchive(node, CONSTANTS.baseCid);
		await expect(store.resolve('a/b/c')).rejects.toBe('path must start with a "/"');
	});
});
