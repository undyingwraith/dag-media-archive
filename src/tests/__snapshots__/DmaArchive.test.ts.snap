// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DmaArchive resolve works 1`] = `
Object {
  "id": "ebb941ae-419e-4848-8664-57597ddb214e",
  "source": Object {
    "codec": "dag-pb",
    "hash": Object {
      "data": Array [
        18,
        32,
        26,
        46,
        122,
        85,
        42,
        12,
        74,
        209,
        84,
        253,
        49,
        89,
        5,
        33,
        29,
        170,
        81,
        241,
        172,
        244,
        68,
        30,
        234,
        69,
        153,
        233,
        192,
        215,
        165,
        228,
        102,
        93,
      ],
      "type": "Buffer",
    },
    "version": 1,
  },
}
`;
